Add in documentation on creating your first PR to Metasploit (#16949)
rapid7/metasploit-framework: docs/metasploit-framework.wiki/Creating-Your-First-PR.md

# Creating Your First PR - An Intro To Git and the PR Process
## Intro

> Congratulations fellow traveler, so your interested in contributing to Metasploit eh? Well welcome aboard, its going to be a fun ride!

Reviewer: ```suggestion
Congratulations fellow traveler, so you're interested in contributing to Metasploit eh? Well welcome aboard, its going to be a fun ride!
```

Author: Resolved in https://github.com/rapid7/metasploit-framework/pull/16949/commits/2bfcac8d8630c70097cd0b029271ec9c25e522d6

You'll learn lots along the way but here are some tips and tricks that should help you get started with making your first PR request
whilst also avoiding some common pitfalls and learning how some of our systems work.

## Initial Steps and Important Notes
The rest of this guide assumes you have already followed the steps at [Setting Up A Developer Environment](https://r-7.co/MSF-DEV) in order to get

> a fork of Metasploit set up and ready to run, and that you have added in your SSH keys, set up Ruby and optionally the PostgreSQL database, and

Reviewer: Can we link to https://docs.github.com/en/authentication/connecting-to-github-with-ssh/adding-a-new-ssh-key-to-your-github-account for SSH keys as you did for "Setting Up A Developer Environment"?

Author: Sure I'll add that in :)

Author: Added in https://github.com/rapid7/metasploit-framework/pull/16949/commits/2bfcac8d8630c70097cd0b029271ec9c25e522d6, let me know if this looks good.

done any custom shortcuts you wish to configure.

## Getting the Latest Version of Metasploit Framework
Before making any new contributions, you will want to sure you are running the latest version of Metasploit Framework.

> To do this run `git checkout master && git fetch && git pull`. This will check out the `master` branch, then fetch all

Reviewer: `git fetch` without another argument is going to fetch the origin. IIRC, the preferred setup is to clone your fork in which case to update from Rapid7, you'd have to have had that setup as another remote and fetch it from there. That's typically the `upstream` remote.

If they followed the step [here](https://docs.metasploit.com/docs/development/get-started/setting-up-a-metasploit-development-environment.html#set-up-your-local-copy-of-the-repository) from the docs you're suggesting then I don't think they'll actually get Rapid7's code.

I'd suggest updating that part to explicitly include the remote and maybe mention what that remote is. Something to the effect of:

```
`git checkout master && git fetch upstream && git pull` where "upstream" is the Rapid7 remote. This can be verified using the command `git remote get-url upstream` and should be `git@github.com:rapid7/metasploit-framework.git
```

Author: Huh interesting it worked for someone who followed that guide however I do fully agree with your statement here and think the extra detail surrounding this is needed. Thanks for pointing this out, I'll update this now.

Author: Added in new explanation in https://github.com/rapid7/metasploit-framework/pull/16949/commits/2bfcac8d8630c70097cd0b029271ec9c25e522d6 using your template, let me know if that works for you.

you intended to modify. In that case maintainers will then have to try find the right combination of changes to implment, which could lead
to your PR being rejected if these changes are too complex.

## Making Sure Your Gems Are Updated
The next step is to make sure you have the latest copy of the Gems that Metasploit Framework depends on. This can be done by running `bundle install`
from the same directory as where the `Gemfile.lock` file is located, which will be in the same folder as wherever you cloned your fork to locally.

Doing this will allow you to make sure that you are running the latest libraries, which will ensure if you do encounter any bugs whilst
developing code, those bugs are not related to out of date Gems being installed, and are therefore potentially legitimate bugs that need fixing.

## Creating a New Branch for Your Code
Once all of this is done, you will want to create a new branch for your code, which can be done by running `git checkout -b <your branch name here>`.
This will snapshot the current branch that you are on, and use that to create a new branch with the name provided. Note that I did say snapshot. This is

> why its important to update the current branch's code to the latest version of Metasploit Framework available prior to running this command,

Reviewer: ```suggestion
why it's important to update the current branch's code to the latest version of Metasploit Framework available prior to running this command,
```

Author: Resolved in https://github.com/rapid7/metasploit-framework/pull/16949/commits/2bfcac8d8630c70097cd0b029271ec9c25e522d6

otherwise the new branch will contain outdated code.

## Adding in Your Changes and Creating Meaningful Commit Messages
Once you have made your code changes, add them using `git add <path to file to add> <optional path to second file to add>`. Note that you can
specify multiple files to add using `git add` at the same time.

To commit these changes locally, use `git commit -m "<commit message here>"`. Note that as a general rule of thumb, commit messages should aim

> to be short whilst also telling readers what was changed in that commit. You generally don't want to create commits that do multiple things at once,

Reviewer: ```suggestion
to be 50 characters or less while telling readers what was changed in that commit. You generally don't want to create commits that do multiple things at once,
```

Author: Resolved in https://github.com/rapid7/metasploit-framework/pull/16949/commits/2bfcac8d8630c70097cd0b029271ec9c25e522d6

instead create a separate commit for each group of items that you are changing, and make sure that the commit message reflects what changed in a general sense.

Note also that maintainers may end up squashing your commits down so that your commit A, B, and C, now become commit D which
contains all of the same changes as commit A, B, and C, but in one commit and with one associated commit message. This is often
done when the code is ready to be landed into Metasploit Framework to help make the commit history easier for people to read.

## Checking for Code Errors
Before code can be accepted into Metasploit Framework, it must also pass our RuboCop and MsfTidy rules. These help ensure that
all contributors are committing code that follows a common set of standards. To check if your code meets our RuboCop standards, 

> from the root of wherever you cloned your fork of Metasploit Framework to on disk, run `rubocop -a <path to your module from current directory>`.

Reviewer: This makes it sound like rubocop will only check, however adding the `-a` argument will cause rubocop to automatically make corrections. Either you should update this to say that rubocop will make corrections, or drop the `-a` argument.

Author: Idea here was to provide more clarity around what the impact of excluding and including the `-a` option was and that omitting it will cause RuboCop to only suggest edits but not actually make them. So yes to some degree this does fall n line with what you read this as so I'm a bit confused why you wanted this edited?

Author: Fixed in 808b221b3d8dfb77530164b196043cd94fa3cfbd

> This will ask RuboCop to check your module and if possible fix any issues that RuboCop is able to fix. Keep running this command and fixing any

Reviewer: If above, you opted to drop the `-a` then here you could update it to be

`Specifying -a will cause RuboCop to check your module and if possible...`

Author: Fixed in 808b221b3d8dfb77530164b196043cd94fa3cfbd

Author: Fixed in 808b221b3d8dfb77530164b196043cd94fa3cfbd

folder located in the root of the Metasploit fork. You can then copy one of these files and use it as the basis for writing
your new documentation for your module.

When writing the information for the documentation, be sure to make sure your installation steps are as clear as possible. Any confusion over
how to set up the target to be exploited will likely result in delays. You will want to put as much detail here as possible.

Additionally any information about caveats, scenarios you have tested, custom options you added in, or quirks you noticed
should also go into this file.

## Checking Documentation Syntax
Once you have written the documentation, you then want to run `toos/dev/msftidy_docs.rb <path to documentation file>`. This will report on any
errors with your documentation file, which you will want to fix before submitting your PR. Notice however that if you get a warning about long lines,
these may be okay to ignore depending on the context. A good example is if a line is long merely because of a URL. Such warnings can be
safely ignored.

## Submitting Your Changes and Opening a PR

> Once you have gone through all of the steps above you should be ready to submit your PR. To submit your PR do `git push origin local-branch:remote-branch` 

Reviewer: This also implies that the `origin` remote is the fork. This confirms my previous comment regarding `git fetch` needing to be clarified.

Author: Added some more explanation to try clear this up a bit more, let me know if this is okay.
